jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceInputStream.java

```diff
@@ -100,23 +100,32 @@ public int available() throws IOException
     @Override
     public void close()
     {
-        // If we have already reached a real EOF or a persistent failure, close is a noop.
-        if (chunk == Content.Chunk.EOF || Content.Chunk.isFailure(chunk, true))
-            return;
-
-        // If we have a chunk here, then it needs to be released
-        if (chunk != null)
+        while (true)
         {
-            chunk.release();
+            // If we have already reached a real EOF or a persistent failure, close is a noop.
+            if (chunk == Content.Chunk.EOF || Content.Chunk.isFailure(chunk, true))
+                return;
 
-            // if the chunk was a last chunk (but not an instanceof EOF), then nothing more to do
-            if (chunk.isLast())
+            // If we have a chunk here, then it needs to be released
+            if (chunk != null)
+            {
+                chunk.release();
+
+                // if the chunk was a last chunk (but not an instanceof EOF), then nothing more to do
+                if (chunk.isLast())
+                    return;
```

Comment:
You should do `chunk = Chunk.next(chunk)` before returning to avoid keeping a ref to a last chunk whose buffer has been released.